### PR TITLE
Fix problems discovered using nextSource() to implement skipValue

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java [hunk 1]
```diff
@@ -564,6 +562,7 @@ public String nextName() throws IOException {
       result = nextQuotedValue(SINGLE_QUOTE_OR_SLASH);
     } else if (p == PEEKED_BUFFERED_NAME) {
       result = peekedString;
+      peekedString = null;
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java:565]: unrelated bug fix. No visible effects, but it was wrong to retain this value when it wasn’t reachable.

File: moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java [hunk 2]
```diff
@@ -1069,9 +1068,14 @@ public BufferedSource nextSource() throws IOException {
       throw new JsonDataException("Expected a value but was " + peek() + " at path " + getPath());
     }
 
+    // Advance the path and clear peeked if we haven't already.
+    if (peeked != PEEKED_NONE) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java:1072]: this check is necessary to avoid double-incrementing the position in cases like unquoted strings.

**Review by swankjesse — COMMENTED** [review]
This passes tests with an alternate `skip()` method.

```
  @Override
  public void skipValue() throws IOException {
    if (failOnUnknown) {
      throw new JsonDataException("Cannot skip unexpected " + peek() + " at " + getPath());
    }
    if (peek() == Token.NAME) {
      skipName();
    } else {
      nextSource();

      // If this element is in an object clear out the key.
      if (stackSize > 1) {
        pathNames[stackSize - 2] = "null";
      }
    }
  }
```

**Review by ZacSweers — APPROVED** [review]
